sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -5768,6 +5798,112 @@ bool _pi_queue::useCopyEngine(bool PreferCopyEngine) const {
+// pointed to by a pair of non-USM pointers A and B.
+// NO_OVERLAP - addresses do not overlap
+// PART_OVERLAP - some addresses overlap
+// FULL_OVERLAP - all addresses of A overlap with those of B
+typedef enum { NO_OVERLAP, PART_OVERLAP, FULL_OVERLAP } OverlapType;
+
+// This function is used for determining overlaps between addresses
+// pointed to by a pair of non-USM pointers A and B.
+// Arguments are:
+// 1. Starting address for A
+// 2. Size of addresses pointed to by A (in bytes)
+// 3. Starting address for B
+// 4. Size of addresses pointed to by B (in bytes)
+// Result is of type OverlapType.
+static OverlapType overlapWithAlreadyImportedPtrs(const void *Src1,
+                                                  size_t Size1,
+                                                  const void *Src2,
+                                                  size_t Size2) {
+  auto Start1 = pi_cast<const char *>(Src1);
+  auto End1 = pi_cast<const char *>(Src1) + Size1;
+  auto Start2 = pi_cast<const char *>(Src2);
+  auto End2 = pi_cast<const char *>(Src2) + Size2;
+  // All addresses in A are also in B
+  if (Start1 >= Start2 && End1 <= End2)
+    return FULL_OVERLAP;
+  // Some addresses in A are in B as well
+  if ((Start1 >= Start2 && Start1 <= End2 && End1 >= Start2) ||
+      (Start2 >= Start1 && Start2 <= End1 && End2 >= Start1))
+    return PART_OVERLAP;
+  // No address overlap between A and B
+  return NO_OVERLAP;
+}
+
+// This function takes a pointer and its associated size as input.
+// If the pointer is a non_USM pointer and USM import is enabled,
+// then, the function tries to import the pointer.
+// An existing map of all imported pointers helps in determining
+// what has already been imported and guides the import process.
+static pi_result importNonUSMHostPtr(const void *Ptr, size_t Size,
+                                     pi_queue Queue) {
+  if (ZeImportHostPtr && ZeUSMImport.Enabled && Ptr != nullptr) {
+    ze_device_handle_t ZeDeviceHandle;
+    ZeStruct<ze_memory_allocation_properties_t> ZeMemoryAllocationProperties;
+    ZE_CALL(zeMemGetAllocProperties,
+            (Queue->Context->ZeContext, Ptr, &ZeMemoryAllocationProperties,
+             &ZeDeviceHandle));
+    // If not shared of any type, we can import the ptr
+    if (ZeMemoryAllocationProperties.type == ZE_MEMORY_TYPE_UNKNOWN) {
+      ze_driver_handle_t driverHandle =
+          Queue->Context->Devices[0]->Platform->ZeDriver;
+      // Temporary data structure to hold all pointers that
+      // will be removed from ImportedHostPtrs.
+      std::vector<const void *> PtrsToErase;
+      // Flag that determines if the incoming pointer will be imported.
+      bool NeedImport = Queue->Context->ImportedHostPtrs.empty();
+
+      // This loop parses through all the imported pointers and
+      // checks for overlaps with incoming pointer
+      for (auto ImportedPtr : Queue->Context->ImportedHostPtrs) {
+        auto Overlap = overlapWithAlreadyImportedPtrs(
```

Comment:
This is O(n^2) complexity. Instead of going through all elements I think you should use an ordered map instead, and only check items that are potentially overlapping, see "lower_bound" and "upper_bound". This should make it down to O(n*logn)